signer_client.go

```diff
@@ -344,3 +373,166 @@ func (s *signerClient) DeriveSharedKey(ctx context.Context,
+		s.TaprootTweak = &signrpc.TaprootTweakDesc{
+			ScriptRoot:   scriptRoot,
+			KeySpendOnly: keySpendOnly,
+		}
+	}
+}
+
+// MuSig2CreateSession creates a new musig session with the key and signers
+// provided.
+func (s *signerClient) MuSig2CreateSession(ctx context.Context,
+	signerLoc *keychain.KeyLocator, signers [][32]byte,
+	opts ...MuSig2SessionOpts) (*input.MuSig2SessionInfo, error) {
+
+	signerBytes := make([][]byte, len(signers))
+	for i, signer := range signers {
+		signerBytes[i] = make([]byte, 32)
+		copy(signerBytes[i], signer[:])
+	}
+
+	req := &signrpc.MuSig2SessionRequest{
+		KeyLoc: &signrpc.KeyLocator{
+			KeyFamily: int32(signerLoc.Family),
+			KeyIndex:  int32(signerLoc.Index),
+		},
+		AllSignerPubkeys: signerBytes,
+	}
+
+	for _, opt := range opts {
+		opt(req)
+	}
+
+	rpcCtx, cancel := context.WithTimeout(ctx, s.timeout)
+	defer cancel()
+
+	rpcCtx = s.signerMac.WithMacaroonAuth(rpcCtx)
+	resp, err := s.client.MuSig2CreateSession(rpcCtx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	combinedKey, err := schnorr.ParsePubKey(resp.CombinedKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse combined key: %v", err)
+	}
+
+	session := &input.MuSig2SessionInfo{
+		CombinedKey:   combinedKey,
+		HaveAllNonces: resp.HaveAllNonces,
+	}
+
+	if len(resp.LocalPublicNonces) != musig2.PubNonceSize {
+		return nil, fmt.Errorf("unexpected local nonce size: %v",
+			len(resp.LocalPublicNonces))
+	}
+	copy(session.PublicNonce[:], resp.LocalPublicNonces)
+
+	if len(resp.SessionId) != 32 {
+		return nil, fmt.Errorf("unexpected session ID length: %v",
+			len(resp.SessionId))
+	}
+
+	copy(session.SessionID[:], resp.SessionId)
+
+	return session, nil
+}
+
+// MuSig2RegisterNonces registers additional public nonces for a musig2 session.
+// It returns a boolean indicating whether we have all of our nonces present.
+func (s *signerClient) MuSig2RegisterNonces(ctx context.Context,
+	sessionID [32]byte, nonces [][musig2.PubNonceSize]byte) (bool, error) {
+
+	req := &signrpc.MuSig2RegisterNoncesRequest{
+		SessionId:               sessionID[:],
+		OtherSignerPublicNonces: noncesToBytes(nonces),
+	}
+
+	rpcCtx, cancel := context.WithTimeout(ctx, s.timeout)
+	defer cancel()
+
+	rpcCtx = s.signerMac.WithMacaroonAuth(rpcCtx)
+	resp, err := s.client.MuSig2RegisterNonces(rpcCtx, req)
+	if err != nil {
+		return false, err
+	}
+
+	return resp.HaveAllNonces, nil
+}
+
+// MuSig2Sign creates a partial signature for the 32 byte SHA256 digest of a
+// message. This can only be called once all public nonces have been created.
+// If the caller will not be responsible for combining the signatures, the
+// cleanup bool should be set.
+func (s *signerClient) MuSig2Sign(ctx context.Context, sessionID [32]byte,
+	message [32]byte, cleanup bool) ([]byte, error) {
+
+	req := &signrpc.MuSig2SignRequest{
+		SessionId:     sessionID[:],
+		MessageDigest: message[:],
+		Cleanup:       cleanup,
+	}
+	rpcCtx, cancel := context.WithTimeout(ctx, s.timeout)
```

Comment:
nit: add newline before to match the other methods?

Reply:
done